discovery_engine_core/core/src/document.rs

```diff
@@ -141,3 +220,165 @@ pub struct UserReacted {
+mod tests {
+    use chrono::NaiveDate;
+    use claim::{assert_matches, assert_none};
+
+    use super::*;
+
+    fn example_url() -> Url {
+        Url::parse("https://example.net").unwrap(/* used only in tests */)
+    }
+
+    impl Default for WebResource {
+        fn default() -> Self {
+            Self {
+                title: String::default(),
+                snippet: String::default(),
+                url: example_url(),
+                display_url: example_url(),
+                date_published: NaiveDate::from_ymd(2022, 1, 1).and_hms(9, 0, 0),
+                provider: None,
+                score: None,
+                rank: 0,
+                country: "en".to_string(),
+                language: "en".to_string(),
+                topic: Topic::Unrecognized,
+            }
+        }
+    }
+
+    impl Default for WebResourceProvider {
+        fn default() -> Self {
+            Self {
+                name: String::default(),
+                thumbnail: None,
+            }
+        }
+    }
+
+    #[test]
+    fn test_web_resource_from_article() {
+        let title = "title".to_string();
+        let clean_url = "https://example.com/".to_string();
+        let excerpt = "summary of the article".to_string();
+        let link = "https://example.com/news/".to_string();
+        let media = "https://example.com/news/image/".to_string();
+        let country = "en".to_string();
+        let language = "en".to_string();
+        let score = Some(0.75);
+        let rank = 10;
+        let topic = Topic::News;
+        let published_date = NaiveDate::from_ymd(2022, 1, 1).and_hms(9, 0, 0);
+
+        let article = Article {
+            id: "id".to_string(),
+            title: title.clone(),
+            score,
+            rank,
+            clean_url: clean_url.clone(),
+            excerpt: excerpt.clone(),
+            link: link.clone(),
+            media: media.clone(),
+            topic: topic.clone(),
+            country: country.clone(),
+            language: language.clone(),
+            published_date,
+        };
+
+        let web_resource: WebResource = article.try_into().unwrap();
+
+        assert_eq!(title, web_resource.title);
+        assert_eq!(clean_url, web_resource.display_url.to_string());
+        assert_eq!(excerpt, web_resource.snippet);
+        assert_eq!(link, web_resource.url.to_string());
+        assert_eq!(country, web_resource.country);
+        assert_eq!(language, web_resource.language);
+        assert_eq!(score, web_resource.score);
+        assert_eq!(rank, web_resource.rank);
+        assert_eq!(topic, web_resource.topic);
+        assert_eq!(published_date, web_resource.date_published);
+
+        let provider = web_resource.provider.unwrap();
+
+        assert_eq!(String::new(), provider.name);
+        assert_eq!(media, provider.thumbnail.unwrap().to_string());
+    }
+
+    #[test]
+    fn test_web_resource_from_article_invalid_links() {
+        let article = Article {
```

Comment:
Instead of writing all of these out by hand and adding a lot of boilerplate for the tests, I think we should load these from the provider fixtures instead I used for the newscatcher tests instead.

Reply:
I refactored this to reduce boilerplate. I don't want to import it from a json because, for example, we add new fields then we will never notice. If we build the instance instead we will have an error and hopefully we will add the relevant tests.
Also we have an error at compile time if the struct changes.
In dart in most places it was loaded from a json and was a bit annoying because you only notice at runtime.